azcausal/estimators/panel/ssnbiclustering.py

```diff
@@ -0,0 +1,721 @@
+from typing import Any, Iterator, List, Optional, Tuple, Union
+import warnings
+import matplotlib.pyplot as plt
+import pandas as pd
+import numpy as np
+from abc import abstractmethod
+from numpy import ndarray, float64
+from sklearn.cluster import SpectralBiclustering
+from sklearn.utils import check_array  # type: ignore
+from tensorly.decomposition import parafac
```

Comment:
I am trying to not add too many dependencies. 

Is this from sklearn an alternative?
https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.non_negative_factorization.html

It seems to be implemented based on 
Fast Hierarchical Alternating Least Squares (Fast HALS).